vmm/src/lib.rs

```diff
@@ -661,18 +666,239 @@ impl Vmm {
+                    let snapshot: Snapshot = serde_json::from_slice(&data).map_err(|e| {
+                        MigratableError::MigrateReceive(anyhow!(
+                            "Error deserialising snapshot: {}",
+                            e
+                        ))
+                    })?;
+
+                    // Create VM
+                    let vm_snapshot = get_vm_snapshot(&snapshot)?;
+                    self.vm_config = Some(Arc::clone(&vm_snapshot.config));
+                    let exit_evt = self.exit_evt.try_clone().map_err(|e| {
+                        MigratableError::MigrateReceive(anyhow!(
+                            "Error cloning exit EventFd: {}",
+                            e
+                        ))
+                    })?;
+                    let reset_evt = self.reset_evt.try_clone().map_err(|e| {
+                        MigratableError::MigrateReceive(anyhow!(
+                            "Error cloning reset EventFd: {}",
+                            e
+                        ))
+                    })?;
+                    let mut vm = Vm::new_from_snapshot(
+                        &snapshot,
+                        exit_evt,
+                        reset_evt,
+                        self.vmm_path.clone(),
+                        None,
+                        false,
+                        &self.seccomp_action,
+                        self.hypervisor.clone(),
+                    )
+                    .map_err(|e| {
+                        MigratableError::MigrateReceive(anyhow!(
+                            "Error creating VM from snapshot: {:?}",
+                            e
+                        ))
+                    })?;
+                    vm.restore(snapshot).map_err(|e| {
+                        Response::error().write_to(&mut socket).ok();
+                        e
+                    })?;
+                    self.vm = Some(vm);
+
+                    Response::ok().write_to(&mut socket)?;
+                }
+                Command::Memory => {
+                    info!("Memory Command Received");
+
+                    if !started {
+                        warn!("Migration not started yet");
+                        Response::error().write_to(&mut socket)?;
+                        continue;
+                    }
+
+                    // Read table
+                    let table = MemoryRangeTable::read_from(&mut socket, req.length())?;
+
+                    // And then read the memory itself
+                    self.vm
+                        .as_mut()
+                        .unwrap()
+                        .receive_memory_regions(&table, &mut socket)
+                        .map_err(|e| {
+                            Response::error().write_to(&mut socket).ok();
+                            e
+                        })?;
+                    Response::ok().write_to(&mut socket)?;
+                }
```

Comment:
Both the memory and state receiving state handling could be moved to separate functions and make `vm_receive_migration` a more readable function.

Reply:
Ok.